src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java

```diff
@@ -19,63 +27,186 @@
     }
 
     @Override
-    public void refreshStore(String key, String slug, String serverId, Stream<BitbucketPullRequest> bbsPullRequests) {
+    public void refreshStore(String projectKey, String slug, String serverId, Stream<BitbucketPullRequest> bbsPullRequests) {
         PullRequestStoreImpl.CacheKey cacheKey =
-                new PullRequestStoreImpl.CacheKey(key, slug, serverId);
-        ConcurrentLinkedQueue<BitbucketPullRequest> pullRequest = pullRequests.get(cacheKey);
-        if (pullRequest != null) {
-            pullRequest.clear();
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        AtomicLong oldestUpdate = pullRequests.getOrDefault(cacheKey, new RepositoryStore(new ConcurrentHashMap<>())).findOldestUpdate();
+
+        //bbspullrequests is ordered from newest to oldest
+        bbsPullRequests.map(pr -> {
+            if (pr.getUpdatedDate() > oldestUpdate.get()) {
+                updatePullRequest(serverId, pr);
+            }
+            return pr.getUpdatedDate();
+        }).filter(updated -> updated < oldestUpdate.get())
+                .findFirst();
+    }
+
+    @Override
+    public void removeClosedPullRequests(long date) {
+       pullRequests.values().forEach(repositoryStore -> repositoryStore.removeClosedPullRequests(date));
+    }
+
+    @Override
+    public boolean hasPRForRepository(String projectKey, String slug, String serverId) {
+        PullRequestStoreImpl.CacheKey key =
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        return pullRequests.getOrDefault(key, new RepositoryStore(new ConcurrentHashMap<>())).hasPR();
+    }
+
+    public static MinimalPullRequest convert(BitbucketPullRequest bbsPR) {
+        return new MinimalPullRequest(bbsPR.getId(), bbsPR.getState(), bbsPR.getFromRef().getDisplayId(),
+                bbsPR.getToRef().getDisplayId(), bbsPR.getUpdatedDate());
+    }
+
+    private static class RepositoryStore {
+
+        private static final MinimalPullRequest CLOSED_PR = new MinimalPullRequest(0, BitbucketPullState.DECLINED,
+                "", "", 0);
+        private ConcurrentMap<String, ConcurrentMap<String, MinimalPullRequest>> pullRequests;
+
+        public RepositoryStore(
+                ConcurrentMap<String, ConcurrentMap<String, MinimalPullRequest>> pullRequests) {
+            this.pullRequests = pullRequests;
+        }
+
+        public void closePullRequest(String fromBranch, String toBranch) {
+            pullRequests.getOrDefault(fromBranch, new ConcurrentHashMap<>()).getOrDefault(toBranch, CLOSED_PR)
+                    .setState(BitbucketPullState.DELETED);
+        }
+
+        public AtomicLong findOldestUpdate() {
+            AtomicLong oldestUpdate = new AtomicLong(0);
+            pullRequests.values().stream().forEach(fromBranch -> {
+                fromBranch.values().forEach(pr -> {
+                    if (oldestUpdate.get() == 0) {
```

Comment:
Rather than having a special case for 0, you can trim this method down by using:
`AtomicLong oldestUpdate = new AtomicLong(Long.MAX_VALUE);`
That way, you don't need the first conditional- the youngest PR will always be older than Long.MAX_VALUE.

Reply:
> the youngest PR will always be **older** than Long.MAX_VALUE

... are you sure? Long.MAX_VALUE for a date is "Sun Aug 17 17:12:55 AEST 292278994" so.. any PR will be _younger_ than that. But we are not looking for the longest, we are looking for the oldest. Mind you I suggested a streamy version above for method. If we go with that this discussion is academic now.